lib/DxilDia/DxcPixDxilStorage.cpp

```diff
@@ -0,0 +1,291 @@
+  }
+
+  DWORD IndexedTypeSizeInBits;
+  IFR(m_pType->GetSizeInBits(&IndexedTypeSizeInBits));
+
+  const unsigned NewOffsetInBits =
+      m_OffsetFromStorageStartInBits + Index * IndexedTypeSizeInBits;
+
+  return CreateDxcPixStorageImpl(
+      m_pDxilDebugInfo,
+      IndexedType,
+      m_pVarInfo,
+      NewOffsetInBits,
+      ppResult);
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilArrayStorage::GetRegisterNumber(
+    _Outptr_result_z_ DWORD* pRegisterNumber)
+{
+  return E_FAIL;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilArrayStorage::GetIsAlive()
+{
+  for (auto OffsetAndRegister : m_pVarInfo->m_ValueLocationMap)
+  {
+    if (OffsetAndRegister.second.m_V != nullptr)
+    {
+      return S_OK;
+    }
+  }
+  return S_FALSE;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilArrayStorage::GetType(
+    _Outptr_result_z_ IDxcPixType** ppType)
+{
+  *ppType = m_pOriginalType;
+  (*ppType)->AddRef();
+  return S_OK;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilStructStorage::AccessField(
+    _In_ LPCWSTR Name,
+    _COM_Outptr_ IDxcPixDxilStorage** ppResult)
+{
+  CComPtr<IDxcPixStructField> Field;
+  IFR(m_pType->GetFieldByName(Name, &Field));
+
+  CComPtr<IDxcPixType> FieldType;
+  IFR(Field->GetType(&FieldType));
+
+  DWORD FieldOffsetInBits;
+  IFR(Field->GetOffsetInBits(&FieldOffsetInBits));
+
+  const unsigned NewOffsetInBits =
+      m_OffsetFromStorageStartInBits + FieldOffsetInBits;
+
+  return CreateDxcPixStorageImpl(
+      m_pDxilDebugInfo,
+      FieldType,
+      m_pVarInfo,
+      NewOffsetInBits,
+      ppResult);
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilStructStorage::Index(
+    _In_ DWORD Index,
+    _COM_Outptr_ IDxcPixDxilStorage** ppResult)
+{
+  return E_FAIL;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilStructStorage::GetRegisterNumber(
+    _Outptr_result_z_ DWORD* pRegisterNumber)
+{
+  return E_FAIL;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilStructStorage::GetIsAlive()
+{
+  for (auto OffsetAndRegister : m_pVarInfo->m_ValueLocationMap)
+  {
+    if (OffsetAndRegister.second.m_V != nullptr)
+    {
+      return S_OK;
+    }
+  }
+  return S_FALSE;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilStructStorage::GetType(
+    _Outptr_result_z_ IDxcPixType** ppType)
+{
+  *ppType = m_pOriginalType;
+  (*ppType)->AddRef();
+  return S_OK;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilScalarStorage::AccessField(
+    _In_ LPCWSTR Name,
+    _COM_Outptr_ IDxcPixDxilStorage** ppResult)
+{
+  return E_FAIL;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilScalarStorage::Index(
+    _In_ DWORD Index,
+    _COM_Outptr_ IDxcPixDxilStorage** ppResult)
+{
+  return E_FAIL;
+}
+
+STDMETHODIMP dxil_debug_info::DxcPixDxilScalarStorage::GetRegisterNumber(
+    _Outptr_result_z_ DWORD* pRegisterNumber)
+{
+  const auto &ValueLocationMap = m_pVarInfo->m_ValueLocationMap;
+  auto RegIt = ValueLocationMap.find(
+      m_OffsetFromStorageStartInBits);
+
+  if (RegIt == ValueLocationMap.end())
+  {
+    return S_FALSE;
```

Comment:
COM rule: all out params must be initialized on failure.

Also, not a fan of S_FALSE. 
If I were a caller of this I would write:
DWORD num = -1; 
if (SUCCEEDED(xxx->GetRegisterNumber(&num))
{
   //I found the register number!
   OutputDebugString( m_regName[num]); //... and crash.
}

Maybe E_UNEXPECTED or something?

Reply:
See above.